### PR TITLE
Add Windows Update Agent history

### PR DESCRIPTION
**Pull request by michoebey**
This PR adds support for parsing the DataStore.edb which holds the tbHistory table of the Windows Update Agent. With this information you can create a timeline of Windows updates that succeeded/failed.

### CODE REVIEW

File: dissect/target/plugins/os/windows/wua_history.py [hunk 1]
```diff
@@ -0,0 +1,879 @@
...
+    def __init__(self, target):
+        super().__init__(target)
+
+        self._datastore = None
+        datastore_path = target.fs.path(self.DATASTORE_PATH)
+        if datastore_path.exists():
+            try:
+                self._datastore = EseDB(datastore_path.open())
+            except Error as e:
+                self.target.log.warning("Error opening Windows DataStore database", exc_info=e)
```
Comment (Horofic) [inline dissect/target/plugins/os/windows/wua_history.py:828]:
```suggestion
        self._datastore = None
        self.update_table = None

        if (path := target.fs.path(self.DATASTORE_PATH)).exists():
            try:
                self._datastore = EseDB(path.open())
                self._update_table = self._datastore.table(self.DATASTORE_UPDATE_TABLE)
            except Exception as e:
                self.target.log.warning("Error opening Windows DataStore database.", exc_info=e)

```

File: dissect/target/plugins/os/windows/wua_history.py [hunk 2]
```diff
@@ -0,0 +1,879 @@
...
+
+    def check_compatible(self) -> None:
+        if not self._datastore:
+            raise UnsupportedPluginError("No Windows DataStore found")
```
Comment (Horofic) [inline dissect/target/plugins/os/windows/wua_history.py:832]:
```suggestion
        if self._datastore or not self._update_table:
            raise UnsupportedPluginError("No Windows Update Agent data found in."))

```

File: dissect/target/plugins/os/windows/wua_history.py [hunk 3]
```diff
@@ -0,0 +1,879 @@
...
+
+    def get_table_records(self) -> Iterator[dict[str, RecordValue]]:
+        try:
+            table = self._datastore.table(self.DATASTORE_UPDATE_TABLE)
+        except KeyError:
+            return None
+
+        for record in table.records():
+            record_data = {}
+            for column in table.columns:
```
Comment (Horofic) [inline dissect/target/plugins/os/windows/wua_history.py:842]:
```suggestion
        for record in self._update_table.records():
            record_data = {}
            for column in self._update_table.columns:

```

File: dissect/target/plugins/os/windows/wua_history.py [hunk 4]
```diff
@@ -0,0 +1,879 @@
...
+                            record_data["result_string"] = WUA_ERROR_CODES_MAP[value][0]
+                            record_data["result_description"] = WUA_ERROR_CODES_MAP[value][1]
+                        else:
+                            record_data["result"] = "Unknown"
+                            record_data["result_string"] = "Unknown error"
+                            record_data["result_description"] = "Unknown error"
+                elif column.name == "Title":
+                    kb = re.search(r"(KB.[0-9]*)", value)
+                    if kb:
+                        record_data["kb"] = kb.group(0)
```
Comment (Horofic) [inline dissect/target/plugins/os/windows/wua_history.py:867]:
```suggestion
                    if kb := re.search(r"KB.[0-9]*", value):
                        record_data["kb"] = kb.group()

```

File: dissect/target/plugins/os/windows/wua_history.py [hunk 5]
```diff
@@ -0,0 +1,879 @@
...
+                record_data[column.name] = value
+            yield record_data
+
+    @export(record=WuaHistoryRecord)
+    def wua_history(self):
```
Comment (Horofic) [inline dissect/target/plugins/os/windows/wua_history.py:872]: Can document the contents of the records that are yielded here? Similar to other plugins, so it shows up in `target-query -f wua_history -h`.  

Also, can you be a little bit more verbose in what you can accomplish as an analist with this plugin here? A good start is using your PR comment as inspiration ;)

File: dissect/target/plugins/os/windows/wua_history.py [hunk 6]
```diff
@@ -0,0 +1,879 @@
...
+        """Return the historical data of the Windows Update Agent from the DataStore.edb file.
+
+        Gives insight into the patch process and level of the system.
+        """
+        for record in self.get_table_records():
+            values = {TBHISTORY_NAME_MAP.get(key, key): value for key, value in record.items()}
```
Comment (Horofic) [inline dissect/target/plugins/os/windows/wua_history.py:878]: Is it intended to return the `key` when the `key` cannot be found in `TBHISTORY_NAME_MAP`?

File: dissect/target/plugins/os/windows/wua_history.py [hunk 7]
```diff
@@ -0,0 +1,879 @@
...
+            yield WuaHistoryRecord(path=self.target.fs.path(self.DATASTORE_PATH), _target=self.target, **values)
```
Comment (Horofic) [inline dissect/target/plugins/os/windows/wua_history.py:879]:
```suggestion
           yield WuaHistoryRecord(
                **values,
                path=self.target.fs.path(self.DATASTORE_PATH),
                _target=self.target,
            )

```

File: .gitignore [hunk 8]
```diff
@@ -9,3 +9,6 @@ __pycache__/
 tests/_docs/api
 tests/_docs/build
 .tox/
+Pipfile
+Pipfile.lock
+settings.json
```
Comment (Horofic) [inline .gitignore:14]: These should be added in your own (global / personal) `.gitignore`

**Review by Horofic — CHANGES_REQUESTED** [review]

File: tests/plugins/os/windows/test_wua_history.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
+from datetime import datetime, timezone
+
+from dissect.target.plugins.os.windows.wua_history import WuaHistoryPlugin
+from tests._utils import absolute_path
+
+
+def test_wua_history_plugin(target_win, fs_win):
```
Comment (Horofic) [inline tests/plugins/os/windows/test_wua_history.py:7]: Please add the correct typehints here.

File: dissect/target/plugins/os/windows/wua_history.py (hunk 5 above)
Comment (Horofic) [inline dissect/target/plugins/os/windows/wua_history.py:872]:
```suggestion
    def wua_history(self) -> Iterator[WuaHistoryRecord]:

```

**Review by Horofic — CHANGES_REQUESTED** [review]

File: dissect/target/plugins/os/windows/wua_history.py (hunk 6 above)
Comment (michoebey) [reply dissect/target/plugins/os/windows/wua_history.py:878]: I grabbed this from the UAL plugin. I assumed it was written this way for a reason so I left it. Can change if desired.

File: dissect/target/plugins/os/windows/wua_history.py (hunk 5 above)
Comment (michoebey) [reply dissect/target/plugins/os/windows/wua_history.py:872]: I'd happily do it, but can you give a example plugin of what you mean? I can't quickly find one where that was done

File: dissect/target/plugins/os/windows/wua_history.py (hunk 1 above)
Comment (michoebey) [reply dissect/target/plugins/os/windows/wua_history.py]: done

File: dissect/target/plugins/os/windows/wua_history.py (hunk 2 above)
Comment (michoebey) [reply dissect/target/plugins/os/windows/wua_history.py]: done

File: dissect/target/plugins/os/windows/wua_history.py (hunk 3 above)
Comment (michoebey) [reply dissect/target/plugins/os/windows/wua_history.py]: done

File: dissect/target/plugins/os/windows/wua_history.py (hunk 4 above)
Comment (michoebey) [reply dissect/target/plugins/os/windows/wua_history.py]: done

File: dissect/target/plugins/os/windows/wua_history.py (hunk 7 above)
Comment (michoebey) [reply dissect/target/plugins/os/windows/wua_history.py]: done

File: tests/plugins/os/windows/test_wua_history.py (hunk 9 above)
Comment (michoebey) [reply tests/plugins/os/windows/test_wua_history.py]: done

File: dissect/target/plugins/os/windows/wua_history.py (hunk 5 above)
Comment (michoebey) [reply dissect/target/plugins/os/windows/wua_history.py]: done

Comment (Horofic) [reply dissect/target/plugins/os/windows/wua_history.py]: You could use this as in example: https://github.com/fox-it/dissect.target/blob/11afdf0e06bec1a25e8bf820bc26564e107a2853/dissect/target/plugins/os/windows/log/schedlgu.py#L124-L142

**Comment by Horofic** [comment]
@michoebey I've committed some suggestions directly in your PR. Can you please check if you agree with them?

File: .gitignore (hunk 8 above)
Comment (Horofic) [inline .gitignore:14]: Can you still remove these edits from the `.gitignore` file.

**Review by Horofic — CHANGES_REQUESTED** [review]

**Review by Horofic — APPROVED** [review]

**Pull request merged**